File: mns-core/src/test/java/cz/fb/manaus/core/dao/AbstractDaoTest.java
package cz.fb.manaus.core.dao;

import cz.fb.manaus.core.model.BetAction;
import cz.fb.manaus.core.model.BetActionType;
import cz.fb.manaus.core.model.Market;
import cz.fb.manaus.core.model.MarketPrices;
import cz.fb.manaus.core.model.Price;
import cz.fb.manaus.core.model.RunnerPrices;
import cz.fb.manaus.core.model.SettledBet;
import cz.fb.manaus.core.model.Side;
import cz.fb.manaus.core.test.AbstractDatabaseTestCase;
import cz.fb.manaus.core.test.CoreTestFactory;
import org.apache.commons.lang3.time.DateUtils;
import org.junit.After;
import org.springframework.beans.factory.annotation.Autowired;

import javax.persistence.EntityManagerFactory;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Date;
import java.util.List;
import java.util.Map;

import static cz.fb.manaus.core.test.CoreTestFactory.newMarket;
import static org.apache.commons.lang3.time.DateUtils.addHours;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

abstract public class AbstractDaoTest extends AbstractDatabaseTestCase {

    public static final String BET_ID = "99999";
    public static final String SPARTA = "Sparta Praha - Banik Ostrava";
    public static final Map<String, String> PROPS = Map.of("reciprocal", "0.9", "lastMatched", "2");
    public static final String MARKET_ID = "33";

    @Autowired
    protected BetActionDao betActionDao;
    @Autowired
    protected MarketDao marketDao;
    @Autowired
    protected MarketPricesDao marketPricesDao;
    @Autowired
    protected SettledBetDao settledBetDao;
    @Autowired
    protected EntityManagerFactory factory;

    @After
    public void cleanUp() {
        marketDao.deleteMarkets(Date.from(Instant.now().plus(5_000, ChronoUnit.DAYS)));
    }

    protected BetAction createAndSaveBetAction(Market market, Date date, Map<String, String> values, String betId) {
        var betAction = new BetAction(BetActionType.PLACE, date, new Price(2d, 3d, Side.LAY), market, CoreTestFactory.DRAW);
        var marketPrices = CoreTestFactory.newMarketPrices(market);
        marketPricesDao.saveOrUpdate(marketPrices);
        betAction.setMarketPrices(marketPrices);
        betAction.setProperties(values);
        betAction.setBetId(betId);
        betActionDao.saveOrUpdate(betAction);
        return betAction;
    }

    protected void createMarketWithSingleAction() {
        var date = Date.from(Instant.now().plus(5, ChronoUnit.HOURS));
        var market = newMarket(MARKET_ID, date, CoreTestFactory.MATCH_ODDS);
        marketDao.saveOrUpdate(market);
        createAndSaveBetAction(market, addHours(date, -1), PROPS, BET_ID);
    }

    protected SettledBet createMarketWithSingleSettledBet() {
        var now = Date.from(Instant.now());
        var market = newMarket(MARKET_ID, now, CoreTestFactory.MATCH_ODDS);
        marketDao.saveOrUpdate(market);
        var actionDate = Date.from(Instant.now().minus(1, ChronoUnit.HOURS));
        var action = createAndSaveBetAction(market, actionDate, PROPS, BET_ID);
        var bet = new SettledBet(action.getSelectionId(), "x", 5, now, action.getPrice());
        bet.setBetAction(action);
        settledBetDao.saveOrUpdate(bet);
        return bet;
    }

    protected void createBet() {
        var current = new Date();
        var market = newMarket();
        var prices = new MarketPrices(1, market, createRPs(2.1d, 2.2d), new Date());
        prices.setTime(DateUtils.addHours(current, -1));
        marketDao.saveOrUpdate(market);
        marketPricesDao.saveOrUpdate(prices);
        betActionDao.saveOrUpdate(createAction(market, prices, "1"));
        var prices2 = new MarketPrices(1, market, createRPs(2.3d, 2.5d), new Date());
        prices2.setTime(current);
        marketPricesDao.saveOrUpdate(prices2);
        betActionDao.saveOrUpdate(createAction(market, prices2, "2"));
        assertThat(marketPricesDao.getPrices(CoreTestFactory.MARKET_ID).size(), is(2));
    }

    private BetAction createAction(Market market, MarketPrices prices, String betId) {
        var place = new BetAction(BetActionType.PLACE, new Date(), new Price(2.2d, 2.1d, Side.LAY),
                market, CoreTestFactory.HOME, betId);
        place.setMarketPrices(prices);
        return place;
    }

    private List<RunnerPrices> createRPs(double price, double lastMatchedPrice) {
        return List.of(
                CoreTestFactory.newBackRP(price, CoreTestFactory.HOME, lastMatchedPrice),
                CoreTestFactory.newBackRP(price, CoreTestFactory.DRAW, lastMatchedPrice),
                CoreTestFactory.newBackRP(price, CoreTestFactory.AWAY, lastMatchedPrice));
    }

}

File: mns-core/src/test/java/cz/fb/manaus/core/dao/AbstractDaoTest.kt
package cz.fb.manaus.core.dao

import cz.fb.manaus.core.model.*
import cz.fb.manaus.core.test.AbstractDatabaseTestCase
import cz.fb.manaus.core.test.CoreTestFactory
import cz.fb.manaus.core.test.CoreTestFactory.newMarket
import org.apache.commons.lang3.time.DateUtils
import org.apache.commons.lang3.time.DateUtils.addHours
import org.hamcrest.CoreMatchers.`is`
import org.junit.After
import org.junit.Assert.assertThat
import org.springframework.beans.factory.annotation.Autowired
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.*
import javax.persistence.EntityManagerFactory

abstract class AbstractDaoTest : AbstractDatabaseTestCase() {

    @Autowired
    protected lateinit var betActionDao: BetActionDao
    @Autowired
    protected lateinit var marketDao: MarketDao
    @Autowired
    protected lateinit var marketPricesDao: MarketPricesDao
    @Autowired
    protected lateinit var settledBetDao: SettledBetDao
    @Autowired
    protected lateinit var factory: EntityManagerFactory

    @After
    fun cleanUp() {
        marketDao.deleteMarkets(Date.from(Instant.now().plus(5000, ChronoUnit.DAYS)))
    }

    protected fun createAndSaveBetAction(market: Market, date: Date, values: Map<String, String>, betId: String?): BetAction {
        val betAction = ModelFactory.newAction(BetActionType.PLACE, date, Price(2.0, 3.0, Side.LAY), market, CoreTestFactory.DRAW)
        val marketPrices = CoreTestFactory.newMarketPrices(market)
        marketPricesDao.saveOrUpdate(marketPrices)
        betAction.marketPrices = marketPrices
        betAction.properties = values
        betAction.betId = betId
        betActionDao.saveOrUpdate(betAction)
        return betAction
    }

    protected fun createMarketWithSingleAction() {
        val date = Date.from(Instant.now().plus(5, ChronoUnit.HOURS))
        val market = newMarket(MARKET_ID, date, CoreTestFactory.MATCH_ODDS)
        marketDao.saveOrUpdate(market)
        createAndSaveBetAction(market, addHours(date, -1), PROPS, BET_ID)
    }

    protected fun createMarketWithSingleSettledBet(): SettledBet {
        val now = Date.from(Instant.now())
        val market = newMarket(MARKET_ID, now, CoreTestFactory.MATCH_ODDS)
        marketDao.saveOrUpdate(market)
        val actionDate = Date.from(Instant.now().minus(1, ChronoUnit.HOURS))
        val action = createAndSaveBetAction(market, actionDate, PROPS, BET_ID)
        val bet = ModelFactory.newSettled(action.selectionId, "x", 5.0, now, action.price)
        bet.betAction = action
        settledBetDao.saveOrUpdate(bet)
        return bet
    }

    protected fun createBet() {
        val current = Date()
        val market = newMarket()
        val prices = ModelFactory.newPrices(1, market, createRPs(2.1, 2.2), Date())
        prices.time = DateUtils.addHours(current, -1)
        marketDao.saveOrUpdate(market)
        marketPricesDao.saveOrUpdate(prices)
        betActionDao.saveOrUpdate(createAction(market, prices, "1"))
        val prices2 = ModelFactory.newPrices(1, market, createRPs(2.3, 2.5), Date())
        prices2.time = current
        marketPricesDao.saveOrUpdate(prices2)
        betActionDao.saveOrUpdate(createAction(market, prices2, "2"))
        assertThat(marketPricesDao.getPrices(CoreTestFactory.MARKET_ID).size, `is`(2))
    }

    private fun createAction(market: Market, prices: MarketPrices, betId: String): BetAction {
        val place = ModelFactory.newAction(BetActionType.PLACE, Date(), Price(2.2, 2.1, Side.LAY),
                market, CoreTestFactory.HOME)
        place.betId = betId
        place.marketPrices = prices
        return place
    }

    private fun createRPs(price: Double, lastMatchedPrice: Double): List<RunnerPrices> {
        return listOf(
                CoreTestFactory.newBackRP(price, CoreTestFactory.HOME, lastMatchedPrice),
                CoreTestFactory.newBackRP(price, CoreTestFactory.DRAW, lastMatchedPrice),
                CoreTestFactory.newBackRP(price, CoreTestFactory.AWAY, lastMatchedPrice))
    }

    companion object {

        const val BET_ID = "99999"
        const val SPARTA = "Sparta Praha - Banik Ostrava"
        val PROPS = mapOf("reciprocal" to "0.9", "lastMatched" to "2")
        const val MARKET_ID = "33"
    }

}
